tensorflow_datasets/object_detection/voc.py

```diff
@@ -119,128 +159,318 @@ def _get_example_objects(annon_filepath):
     # pytype: enable=attribute-error
 
 
+class VocPath:
+  def __init__(self, filename, relative_path):
+    self.filename = filename
+    self.relative_path = relative_path
+
+  def resolve(self, paths, example_id):
+    return os.path.join(
+        paths[self.filename], os.path.normpath(self.relative_path)
+    ).format(example_id)
+
+
+class DetectionExampleBuilder:
+  """A helper class to build VOC object detection examples."""
+
+  def __init__(self, year=None, filename=None, image_path=None, annotation_path=None):
+    if image_path is None:
+      image_path = VocPath(filename,
+          "VOCdevkit/VOC{}/JPEGImages/{{}}.jpg".format(year))
+    if annotation_path is None:
+      annotation_path = VocPath(filename,
+          "VOCdevkit/VOC{}/Annotations/{{}}.xml".format(year))
+    self.image_path = image_path
+    self.annotation_path = annotation_path
+
+  def build(self, paths, example_id):
+    if self.annotation_path:
+      annotation_filepath = self.annotation_path.resolve(paths, example_id)
+      objects = list(_get_detection_objects(annotation_filepath))
+      # Use set() to remove duplicates.
+      labels = sorted(set(obj["label"] for obj in objects))
+      labels_no_difficult = sorted(set(
+          obj["label"] for obj in objects if obj["is_difficult"] == 0
+      ))
+    else:  # If the annotation path is False there are no annotations.
+      objects = []
+      labels = []
+      labels_no_difficult = []
+    return {
+        "image": self.image_path.resolve(paths, example_id),
+        "image/filename": example_id + ".jpg",
+        "objects": objects,
+        "labels": labels,
+        "labels_no_difficult": labels_no_difficult,
+    }
+
+class SegmentationExampleBuilder(DetectionExampleBuilder):
+  """A helper class to build VOC segmentation examples."""
+
+  def __init__(self, year=None, filename=None, image_path=None, annotation_path=None,
+      class_mask_path=None, instance_mask_path=None):
+    super().__init__(year, filename, image_path, annotation_path)
+    if class_mask_path is None:
+      class_mask_path = VocPath(filename,
+          "VOCdevkit/VOC{}/SegmentationClass/{{}}.png".format(year))
+    if instance_mask_path is None:
+      instance_mask_path = VocPath(filename,
+          "VOCdevkit/VOC{}/SegmentationObject/{{}}.png".format(year))
+    self.class_mask_path = class_mask_path
+    self.instance_mask_path = instance_mask_path
+
+  @staticmethod
+  def _load_png_segmentation_mask(path):
+    """Load a '.png' segmentation mask, ignoring any colour map."""
+    with tf.io.gfile.GFile(path, "rb") as fp:
+      image = tfds.core.lazy_imports.PIL_Image.open(fp)
+    return np.expand_dims(np.array(image, dtype=np.uint8), axis=-1)
```

Comment:
@tanzheny

Reply:
@tanzhenyu

Reply:
@qlzh727 It really seems that nobody want to open a PR to TF https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/png/png_io.cc :wink:

Reply:
TBH, I am not competent enough to update the PNG ops in TF. :)

Reply:
> TBH, I am not competent enough to update the PNG ops in TF. :)

You need only to add an argument to the function to make this condition optional:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/png/png_io.cc#L285-L286